Doc/move irep docs from util to irep

owen-mc-diffblue opened the pull request.

Commit:
Move documentation of irept to be above irept

It was in the documentation for the util folder.

Commit:
Move dstringt documentation to above dstringt

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+/// The `irept::pretty` function outputs the contents of an `irept` directly
+/// and can be used to understand and debug problems with `irept`s.
+///
+/// On their own `irept`s do not “mean” anything; they are effectively
```
Mix of smart and normal quotes

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+///   is used for named children, i.e.  subexpressions, parameters, etc.
+///
+/// * `comments`:   Another map from `irep_namet` to `irept` which is used
+///   for annotations and other ‘non-semantic’ information
```
Probably worth mentioning the caveat about these not being taken into account by the default `operator==` here

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+/// On their own `irept`s do not “mean” anything; they are effectively
+/// generic tree nodes. Their interpretation depends on the contents of
+/// result of the `id` function (the `data`) field. `util/irep_ids.txt`
+/// contains the complete list of `id` values. During the build process it
```
Nah, ireps that use other strings (e.g. `irept("Hello world")` are not common but are definitely present in the codebase

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+/// with the architecture of the code.
+///
+/// Many of the key descendant of `exprt` are declared in `std_expr.h`. All
+/// expressions have a named subfield / annotation which gives the type of
```
`named subexpression with ID "type"` would probably be clearer

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+/// expressions have a named subfield / annotation which gives the type of
+/// the expression (slightly simplified from C/C++ as `unsignedbv_typet`,
+/// `signedbv_typet`, `floatbv_typet`, etc.). All type conversions are
+/// explicit with an expression with `id() == ID_typecast` and an ‘interface
```
"interface class" is not used anywhere else, just say "i.e. a `typecast_exprt`" instead

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+/// using the `get_identifier` accessor function.
+///
+/// `codet` inherits from `exprt` and is defined in `std_code.h`. It
+/// represents executable code; statements in C rather than expressions. In
```
...a C-like language...

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+/// the front-end there are versions of these that hold whole code blocks,
+/// but in goto-programs these have been flattened so that each `irept`
+/// represents one sequence point (almost one line of code / one
+/// semi-colon). The most common descendants of `codet` are `code_assignt`
```
is `code_assignt`

smowton reviewed (commented):
Might as well update this while we're doing this

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 105e297b388e75dab6ebfaccbe329aacd0f65f64).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80008898

peterschrammel commented on src/util/dstring.h:
```diff
@@ -16,8 +16,19 @@ Author: Daniel Kroening, kroening@kroening.com
+/// of the same string you only have to store the whole string once, which
+/// saves space.
+///
+/// `irep_idt` and `irep_namet` are typedef-ed to `dstringt` in irep.h unless
```
unless what?

peterschrammel commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+///
+/// On their own `irept`s do not “mean” anything; they are effectively
+/// generic tree nodes. Their interpretation depends on the contents of
+/// result of the `id` function (the `data`) field. `util/irep_ids.txt`
```
This is `irep_ids.def` now.

peterschrammel commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+/// * `sub`:   A vector of `irept` which is used to store ordered but
+///   unnamed children.
+///
+/// The `irept::pretty` function outputs the contents of an `irept` directly
```
maybe say that it prints the explicit tree structure instead of 'directly'

peterschrammel commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+/// * `data`:   A string[^2], which is returned when the `id()` function is
+///   used.
+///
+/// * `named_sub`:   A map from `irep_namet` (a string) to `irept`. This
```
Use as many \ref as possible

owen-mc-diffblue replied:
Annoyingly, doxygen sometimes can't get the \ref to work if the name is in backticks, but sometimes it can

peterschrammel commented on src/util/irep.h:
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
+/// so a common pattern is to cast the `codet` to an assignment and then
+/// recurse on the expression on either side.
+///
+/// [^1]: Or references, if reference counted data sharing is enabled. It is
```
Maybe inline these footnotes.

Commit:
Address review comments

Commit:
Add \ref in lots of places

It often doesn't work when backticks are used, so I've removed them.
They don't affect how the text appears that much when I look at the
html.

owen-mc-diffblue commented:
I have addressed all the review comments

Commit:
Wrap lines properly

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ae66e6517acd37db581202a0316d939e960146f8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80350969

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 6b0dfb65fb758ba394daa844c58a0b10eae67b9e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80355880

owen-mc-diffblue commented:
@peterschrammel @smowton Please re-review.

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,78 @@ inline const std::string &name2string(const irep_namet &n)
+///
+/// * \ref irept::dt::comments : Another map from `irep_namet` to \ref irept
+///   which is used for annotations and other ‘non-semantic’ information. Note
+///   that this map is ignore by the default \ref operator==.
```
ignore*d*

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,78 @@ inline const std::string &name2string(const irep_namet &n)
+/// result of the \ref id() function, i.e. the `data` field. `util/irep_ids.def`
+/// contains a list of `id` values. During the build process it is used
+/// to generate `util/irep_ids.h` which gives constants for each id
+/// (named `ID_`). You can also make `irep_id`s which do not come from
```
irep_idt?

smowton commented on src/util/irep.h:
```diff
@@ -82,8 +82,78 @@ inline const std::string &name2string(const irep_namet &n)
+/// type of the expression (slightly simplified from C/C++ as \ref
+/// unsignedbv_typet, \ref signedbv_typet, \ref floatbv_typet, etc.). All type
+/// conversions are explicit with an expression with `id() == ID_typecast`
+/// and a \ref typecast_exprt. One key descendant of \ref exprt is \ref
```
The typecast_exprt *is* the expression with `id() == ID_typecast`

Commit:
Address review comments #2

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 08558726bc3de3d50ca38aba91a204f6d4091a24).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80546022
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##
